### DISCUSSION TITLE
Slots work differently than I'd anticipate

### DISCUSSION THREAD

**Discussion by idontknowanythingyouknow** [peter-kish/gloot#127] [Q&A] [answered]
Hey, I really love this tool, but I'm a little confused about slots. My thinking would be that slots would work like item slots in Diablo (i.e. an item that is equipped doesn't show in the inventory, an item can only be in one slot at a time). 

However an item in a slot in GLoot still takes up space in its inventory, and can be in several slots at the same time, and you can't drag and drop from an item slot to a grid inventory. I considered using a grid inventory to fill this role, but it has issues with items with different sizes. 

Am I misunderstanding the sentiment behind slots? Is there another design pattern I should use to fulfill the Diablo style inventory?

**Comment by peter-kish** [comment, marked as answer]
I think you're making some very good points here.

When it comes to equipping items into slots, yeah, it is designed so that the item stays inside the inventory even after it's equipped (Deus Ex style). But you're right by pointing out that not all games do that and it would probably be a good idea to make this `ItemSlot` "feature" optional. I'll see if I can squeeze this into version 2.3.8, but I'm also working on version 3.0.0 which should include a lot of changes to the interface and overall design, and it could also be a good place to introduce such a feature.

As for the case where multiple slots can hold the same item - I think it's something I didn't think through very well when I designed it. I guess it's best to consider it a bug which should be fixed in a future 2.3.x version. The same could be said for dragging items from a slot to a grid inventory - I don't see any reason why this shouldn't be possible.

Unfortunately, I'm not sure how Diablo-style slots could be implemented in the current version. You can try experimenting with grid inventories of various sizes, but I'm afraid this could require some additional workaround logic and/or result in additional bugs :(

**Reply by tomek-illu** [reply]
On the topic of item_slots, can they resize the item sprite?

**Reply by peter-kish** [reply]
Scaling `CtrlItemSlot`s is currently not possible, though it would be nice to have that option. I opened an issue for that one as well: https://github.com/peter-kish/gloot/issues/131. It shouldn't be too much work to implement.

**Comment by peter-kish** [comment]
Opened a few issues related to this:
* https://github.com/peter-kish/gloot/issues/128
* https://github.com/peter-kish/gloot/issues/129
* https://github.com/peter-kish/gloot/issues/130

**Comment by Naeta2** [comment]
Am a little late, but wanted to talk about a method that looks close to diablo's inventory and work with gloot, with just a few lines of code to overwrite
I used the same equipment method as kenshi, where your equipment slots are grid inventories the size of the biggest item you can put in it. It's a system that looks a bit junky when you are not used to but with time I grew to like it. In kenshi it does look like [this](https://lofigames.com/wp-content/uploads/weapons.jpg)

In my little game, the first itteration of this inventory design looked like this :
<details>
  <summary>Image</summary>
  
![image](https://github.com/peter-kish/gloot/assets/46900232/66583999-5015-482a-8901-8b4b0c7e8899)

  
</details>

as for the code, there is just a new custom inventory grid that inherits from inventory_grid.gd with those changes : 

<details>
  <summary>Code</summary>
  

  ```
@tool

extends InventoryGrid

class_name CustomInventoryGrid

@export var item_slot : String
@export var is_an_equipment_slot : bool

func can_hold_item(item : InventoryItem) -> bool :
	if item.get_property("slot_type", null) == item_slot :
		return true
	else :
		return false

func can_add_item(item: InventoryItem) -> bool:
	if item == null || has_item(item):
		return false
		
	if !can_hold_item(item):
		return false
		
	if !_constraint_manager.has_space_for(item):
		return false
		
	if is_an_equipment_slot && get_item_count() > 0 :
		return false
		
	return true
  ```
  
</details>

As i said, this type of inventory looks a bit junky but I think it is a convenient solution with GLoot's current version, as it doesn't require much code to overwrite

**Comment by peter-kish** [comment]
**Update:** [Version 2.4.0](https://github.com/peter-kish/gloot/releases/tag/v2.4.0) changes the logic of item slots and they should work more intuitively now.
More information [here](https://github.com/peter-kish/gloot/discussions/144).
